### DISCUSSION TITLE
Token retrieval takes too long

### DISCUSSION THREAD

**Discussion by matejvasek** [Azure/acr#692] [Q&A]
While investigating  [performance issue](https://github.com/buildpacks/lifecycle/issues/1007) of [kpack](https://github.com/pivotal/kpack/)/[lifecycle](https://github.com/buildpacks/lifecycle) I found out that most of the time is spend by auth token retrieval.
It seems to take around two minutes. Any ideas what could be wrong?

Stack:
```
net/http.(*Client).Do
github.com/Azure/go-autorest/autorest/adal.retryForIMDS
github.com/Azure/go-autorest/autorest/adal.(*ServicePrincipalToken).refreshInternal
github.com/Azure/go-autorest/autorest/adal.(*ServicePrincipalToken).EnsureFreshWithContext
github.com/chrismellard/docker-credential-acr-env/pkg/registry.GetRegistryRefreshTokenFromAADExchange
github.com/chrismellard/docker-credential-acr-env/pkg/credhelper.ACRCredHelper.Get
github.com/google/go-containerregistry/pkg/authn.wrapper.Resolve
github.com/buildpacks/lifecycle/auth.ReferenceForRepoName
github.com/buildpacks/lifecycle/auth.buildAuthConfigs
github.com/buildpacks/lifecycle/auth.NewResolvedKeychain
github.com/buildpacks/lifecycle/auth.DefaultKeychain
main.(*exportCmd).Privileges
github.com/buildpacks/lifecycle/cmd/lifecycle/cli.Run
main.main
runtime.main
```

**Comment by matejvasek** [comment]
cc @smile37773 @zhoufenqin

**Comment by matejvasek** [comment]
@smile37773 please provide details if possible.

**Comment by smile37773** [comment]
I use KPack to build images, see more details in https://github.com/pivotal/kpack/blob/main/docs/tutorial.md.
When I use acr to host my image, it takes 6 minutes to complete the build. But when I redirect the image to docker hub, it takes 30s to complete the build.

@matejvasek found the most of the time is spend by auth token retrieval.

The credential we used is the basic-auth secret in k8s.
![image](https://user-images.githubusercontent.com/34502364/234556123-1f40dd86-5039-4e20-8999-3f0a4247c6a8.png)

**Comment by matejvasek** [comment]
There are some updates:
https://github.com/buildpacks/lifecycle/issues/1007#issuecomment-1551273828
https://github.com/buildpacks/lifecycle/issues/1007#issuecomment-1551292693
